src/app/template/index.ts

```diff
@@ -76,135 +97,120 @@ export class ChartTemplateBuilder {
+                  )) {
+                    if (
+                      ps.kind == "chart-element" &&
+                      Prototypes.isType(ps.object.classID, "plot-segment")
+                    ) {
+                      groupBy = (ps.chartElement as Specification.PlotSegment)
+                        .groupBy;
+                      break; // TODO: for now, we assume it's the first one
+                    }
                   }
                 }
               }
-              if (scaleInference.slotName != null) {
-                this.addSlot(
-                  table,
-                  scaleInference.slotName,
-                  scaleInference.slotKind
-                );
-              }
-            }
-            break;
-          case "axis":
-            {
-              const axisInference = item as Specification.Template.Axis;
-              if (axisInference.slotName != null) {
-                this.addSlot(
-                  table,
-                  axisInference.slotName,
-                  axisInference.slotKind
-                );
-              }
-            }
-            break;
-          case "order":
-            {
-              const orderInference = item as Specification.Template.Order;
-              if (orderInference.slotName != null) {
-                this.addSlot(
-                  table,
-                  orderInference.slotName,
-                  orderInference.slotKind
-                );
-              }
             }
-            break;
-          case "slot-list":
-            {
-              const slotListInference = item as Specification.Template.SlotList;
-              for (const slot of slotListInference.slots) {
-                this.addSlot(table, slot.slotName, slot.slotKind);
-              }
-            }
-            break;
+          }
+          inference.scale.expressions = Array.from(expressions);
+          if (!inference.dataSource) {
+            inference.dataSource = {
+              table,
+              groupBy
+            };
+          }
+        }
+        if (inference.expression) {
+          this.addColumnsFromExpression(
+            inference.dataSource.table,
+            inference.expression.expression
+          );
         }
+        template.inference.push(inference);
       }
     }
-    if (params && params.properties && params.properties.length > 0) {
-      template.properties[objectClass.object._id] = params.properties;
+    if (params && params.properties) {
       for (const property of params.properties) {
+        // Make a default display name
         let pn = "";
-        if (property.mode == "attribute") {
-          pn = property.attribute;
-        } else {
-          pn = property.property;
-          if (property.fields) {
-            if (typeof property.fields == "string") {
-              pn += "." + property.fields;
-            } else {
-              pn += "." + property.fields.join(".");
+        if (property.target.property) {
```

Comment:
Seems like it would make sense to make this a variable instead of doing `property.target.property` everywhere.